File: backend/app/models/reporting.py
# =============================================================================
# Stratum AI - Automated Reporting Database Models
# =============================================================================
"""
Database models for automated reporting system.

Models:
- ReportTemplate: Configurable report definitions
- ScheduledReport: Report scheduling configuration
- ReportExecution: History of report runs
- ReportDelivery: Delivery tracking for each channel
"""

import enum
from datetime import date, datetime, timezone
from typing import Optional
from uuid import uuid4

import sqlalchemy as sa
from sqlalchemy import (
    BigInteger,
    Boolean,
    Column,
    Date,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID
from sqlalchemy.orm import relationship

from app.db.base_class import Base, StrEnumType

# =============================================================================
# Enums
# =============================================================================


class ReportType(str, enum.Enum):
    """Types of reports available."""

    CAMPAIGN_PERFORMANCE = "campaign_performance"
    ATTRIBUTION_SUMMARY = "attribution_summary"
    PACING_STATUS = "pacing_status"
    PROFIT_ROAS = "profit_roas"
    PIPELINE_METRICS = "pipeline_metrics"
    EXECUTIVE_SUMMARY = "executive_summary"
    CUSTOM = "custom"


class ReportFormat(str, enum.Enum):
    """Output formats for reports."""

    PDF = "pdf"
    CSV = "csv"
    EXCEL = "excel"
    JSON = "json"
    HTML = "html"


class ScheduleFrequency(str, enum.Enum):
    """How often to run scheduled reports."""

    DAILY = "daily"
    WEEKLY = "weekly"
    BIWEEKLY = "biweekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    CUSTOM = "custom"  # Uses cron expression


class DeliveryChannel(str, enum.Enum):
    """Delivery channels for reports."""

    EMAIL = "email"
    SLACK = "slack"
    TEAMS = "teams"
    WHATSAPP = "whatsapp"
    WEBHOOK = "webhook"
    S3 = "s3"


class ExecutionStatus(str, enum.Enum):
    """Status of report execution."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class DeliveryStatus(str, enum.Enum):
    """Status of report delivery."""

    PENDING = "pending"
    SENT = "sent"
    DELIVERED = "delivered"
    FAILED = "failed"
    BOUNCED = "bounced"


# =============================================================================
# Report Template
# =============================================================================


class ReportTemplate(Base):
    """
    Configurable report template defining what data to include.
    """

    __tablename__ = "report_templates"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)

    # Template identification
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    report_type = Column(StrEnumType(ReportType), nullable=False)

    # Status
    is_active = Column(Boolean, default=True, nullable=False)
    is_system = Column(Boolean, default=False, nullable=False)  # Built-in templates

    # Configuration
    config = Column(JSONB, nullable=False, default={})
    """
    Config structure varies by report_type:
    {
        "metrics": ["spend", "revenue", "roas", "conversions"],
        "dimensions": ["campaign", "platform", "date"],
        "filters": {"platforms": ["meta", "google"], "min_spend": 100},
        "date_range": {"type": "last_30_days"} or {"start": "2026-01-01", "end": "2026-01-31"},
        "comparison": {"enabled": true, "period": "previous_period"},
        "sections": ["summary", "charts", "detailed_table"],
        "branding": {"logo_url": "...", "primary_color": "#..."}
    }
    """

    # Output settings
    default_format = Column(
        StrEnumType(ReportFormat), default=ReportFormat.PDF, nullable=False
    )
    available_formats = Column(ARRAY(String), default=["pdf", "csv"], nullable=False)

    # Styling
    template_html = Column(Text, nullable=True)  # Custom HTML template
    chart_config = Column(JSONB, nullable=True)  # Chart configurations

    # Metadata
    created_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )
    last_modified_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    created_by = relationship("User", foreign_keys=[created_by_user_id])
    last_modified_by = relationship("User", foreign_keys=[last_modified_by_user_id])
    schedules = relationship(
        "ScheduledReport", back_populates="template", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index("ix_report_template_type", "report_type"),
        # was tenant-scoped; now global
        UniqueConstraint("name", name="uq_report_template_name"),
    )


# =============================================================================
# Scheduled Report
# =============================================================================


class ScheduledReport(Base):
    """
    Configuration for scheduled report generation.
    """

    __tablename__ = "scheduled_reports"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    template_id = Column(
        UUID(as_uuid=True),
        ForeignKey("report_templates.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Schedule identification
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # Status
    is_active = Column(Boolean, default=True, nullable=False)
    is_paused = Column(Boolean, default=False, nullable=False)

    # Schedule configuration
    frequency = Column(StrEnumType(ScheduleFrequency), nullable=False)
    cron_expression = Column(String(100), nullable=True)  # For custom frequency
    timezone = Column(String(50), default="UTC", nullable=False)

    # For weekly: which day (0=Monday, 6=Sunday)
    day_of_week = Column(Integer, nullable=True)
    # For monthly: which day (1-31, or -1 for last day)
    day_of_month = Column(Integer, nullable=True)
    # Time to run (hour in 24h format)
    hour = Column(Integer, default=8, nullable=False)
    minute = Column(Integer, default=0, nullable=False)

    # Report configuration overrides
    format_override = Column(StrEnumType(ReportFormat), nullable=True)
    config_override = Column(JSONB, nullable=True)  # Overrides template config

    # Date range for report (relative)
    date_range_type = Column(String(50), default="last_30_days", nullable=False)
    """
    Options: yesterday, last_7_days, last_30_days, last_month,
             month_to_date, quarter_to_date, year_to_date, custom
    """

    # Delivery configuration
    delivery_channels = Column(ARRAY(String), default=["email"], nullable=False)
    delivery_config = Column(JSONB, nullable=False, default={})
    """
    {
        "email": {
            "recipients": ["user@example.com"],
            "cc": [],
            "subject_template": "{{report_name}} - {{date_range}}",
            "body_template": "Please find attached..."
        },
        "slack": {
            "channel": "#reports",
            "webhook_url": "https://hooks.slack.com/..."
        },
        "teams": {
            "webhook_url": "https://outlook.office.com/webhook/..."
        }
    }
    """

    # Execution tracking
    last_run_at = Column(DateTime(timezone=True), nullable=True)
    last_run_status = Column(StrEnumType(ExecutionStatus), nullable=True)
    next_run_at = Column(DateTime(timezone=True), nullable=True)
    run_count = Column(Integer, default=0, nullable=False)
    failure_count = Column(Integer, default=0, nullable=False)

    # Metadata
    created_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    template = relationship("ReportTemplate", back_populates="schedules")
    created_by = relationship("User", foreign_keys=[created_by_user_id])
    executions = relationship(
        "ReportExecution", back_populates="schedule", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index(
            "ix_scheduled_report_next_run",
            "next_run_at",
            postgresql_where=sa.text("is_active = true AND is_paused = false"),
        ),
        Index("ix_scheduled_report_template", "template_id"),
    )


# =============================================================================
# Report Execution
# =============================================================================


class ReportExecution(Base):
    """
    History of report generation runs.
    """

    __tablename__ = "report_executions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    template_id = Column(
        UUID(as_uuid=True),
        ForeignKey("report_templates.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    schedule_id = Column(
        UUID(as_uuid=True),
        ForeignKey("scheduled_reports.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    # Execution details
    execution_type = Column(String(50), nullable=False)  # scheduled, manual, api
    status = Column(
        StrEnumType(ExecutionStatus), nullable=False, default=ExecutionStatus.PENDING
    )

    # Timing
    started_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    completed_at = Column(DateTime(timezone=True), nullable=True)
    duration_seconds = Column(Float, nullable=True)

    # Report parameters
    report_type = Column(StrEnumType(ReportType), nullable=False)
    format = Column(StrEnumType(ReportFormat), nullable=False)
    date_range_start = Column(Date, nullable=False)
    date_range_end = Column(Date, nullable=False)
    config_used = Column(JSONB, nullable=True)  # Snapshot of config at execution time

    # Output
    file_path = Column(String(500), nullable=True)  # Path to generated file
    file_size_bytes = Column(BigInteger, nullable=True)
    file_url = Column(Text, nullable=True)  # Signed URL for download
    file_url_expires_at = Column(DateTime(timezone=True), nullable=True)

    # Report data summary
    row_count = Column(Integer, nullable=True)
    metrics_summary = Column(JSONB, nullable=True)  # Quick summary of key metrics

    # Error tracking
    error_message = Column(Text, nullable=True)
    error_details = Column(JSONB, nullable=True)
    retry_count = Column(Integer, default=0, nullable=False)

    # Triggered by
    triggered_by_user_id = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )

    # Relationships
    template = relationship("ReportTemplate", foreign_keys=[template_id])
    schedule = relationship("ScheduledReport", back_populates="executions")
    triggered_by = relationship("User", foreign_keys=[triggered_by_user_id])
    deliveries = relationship(
        "ReportDelivery", back_populates="execution", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index("ix_report_execution_started", "started_at"),
        Index("ix_report_execution_status", "status"),
        Index("ix_report_execution_schedule", "schedule_id"),
    )


# =============================================================================
# Report Delivery
# =============================================================================


class ReportDelivery(Base):
    """
    Tracks delivery of reports to each channel.
    """

    __tablename__ = "report_deliveries"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    execution_id = Column(
        UUID(as_uuid=True),
        ForeignKey("report_executions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Delivery details
    channel = Column(StrEnumType(DeliveryChannel), nullable=False)
    status = Column(
        StrEnumType(DeliveryStatus), nullable=False, default=DeliveryStatus.PENDING
    )

    # Recipient info
    recipient = Column(String(500), nullable=False)  # Email, channel name, webhook URL
    recipient_type = Column(String(50), nullable=True)  # user, group, webhook

    # Timing
    queued_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    sent_at = Column(DateTime(timezone=True), nullable=True)
    delivered_at = Column(DateTime(timezone=True), nullable=True)

    # Delivery metadata
    message_id = Column(
        String(255), nullable=True
    )  # External message ID (email, Slack ts)
    delivery_response = Column(JSONB, nullable=True)  # Response from delivery service

    # Error tracking
    error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, default=0, nullable=False)
    last_retry_at = Column(DateTime(timezone=True), nullable=True)

    # Relationships
    execution = relationship("ReportExecution", back_populates="deliveries")

    __table_args__ = (
        Index("ix_report_delivery_execution", "execution_id"),
        Index("ix_report_delivery_status", "status"),
        Index("ix_report_delivery_channel", "channel"),
    )


# =============================================================================
# Delivery Channel Configuration
# =============================================================================


class DeliveryChannelConfig(Base):
    """
    Tenant-level configuration for delivery channels.
    """

    __tablename__ = "delivery_channel_configs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)

    # Channel
    channel = Column(StrEnumType(DeliveryChannel), nullable=False)
    name = Column(String(255), nullable=False)  # Friendly name

    # Status
    is_active = Column(Boolean, default=True, nullable=False)
    is_verified = Column(Boolean, default=False, nullable=False)

    # Configuration (encrypted sensitive data)
    config = Column(JSONB, nullable=False)
    """
    Email:
    {
        "smtp_host": "smtp.gmail.com",
        "smtp_port": 587,
        "smtp_user": "...",
        "smtp_password_enc": "...",
        "from_email": "reports@company.com",
        "from_name": "Stratum Reports"
    }

    Slack:
    {
        "webhook_url_enc": "...",
        "bot_token_enc": "...",
        "default_channel": "#reports"
    }

    Teams:
    {
        "webhook_url_enc": "..."
    }

    S3:
    {
        "bucket": "my-reports",
        "prefix": "stratum/",
        "aws_access_key_enc": "...",
        "aws_secret_key_enc": "...",
        "region": "us-east-1"
    }
    """

    # Timestamps
    created_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships

    __table_args__ = (
        # was tenant-scoped; now global
        UniqueConstraint("channel", "name", name="uq_delivery_config"),
    )

File: frontend/src/api/reporting.ts
/**
 * ADs Growth System - Automated Reporting API
 *
 * Handles report templates, scheduling, generation, and delivery.
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse } from './client'

// =============================================================================
// Types
// =============================================================================

export type ReportType = 'campaign_performance' | 'attribution_summary' | 'pacing_status' | 'profit_roas' | 'pipeline_metrics' | 'executive_summary' | 'custom'
export type ReportFormat = 'pdf' | 'csv' | 'excel' | 'json' | 'html'
export type ScheduleFrequency = 'daily' | 'weekly' | 'biweekly' | 'monthly' | 'quarterly' | 'custom'
export type DeliveryChannel = 'email' | 'slack' | 'teams' | 'whatsapp' | 'webhook' | 's3'
export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
export type DeliveryStatus = 'pending' | 'sent' | 'delivered' | 'failed' | 'bounced'

export interface ReportTemplate {
  id: string
  tenantId: number
  name: string
  description?: string
  reportType: ReportType
  config: ReportConfig
  defaultFormat: ReportFormat
  availableFormats: ReportFormat[]
  templateHtml?: string
  chartConfig?: Record<string, any>
  isActive: boolean
  isSystem: boolean
  createdAt: string
  updatedAt: string
}

export interface ReportConfig {
  metrics?: string[]
  dimensions?: string[]
  filters?: Record<string, any>
  dateRange?: { type: string; start?: string; end?: string }
  comparison?: { enabled: boolean; period: string }
  sections?: string[]
  branding?: { logoUrl?: string; primaryColor?: string }
}

export interface ScheduledReport {
  id: string
  tenantId: number
  templateId: string
  template?: ReportTemplate
  name: string
  description?: string
  frequency: ScheduleFrequency
  timezone: string
  dayOfWeek?: number
  dayOfMonth?: number
  hour: number
  minute: number
  cronExpression?: string
  formatOverride?: ReportFormat
  configOverride?: Record<string, any>
  dateRangeType: string
  deliveryChannels: DeliveryChannel[]
  deliveryConfig: DeliveryConfig
  isActive: boolean
  isPaused: boolean
  lastRunAt?: string
  lastRunStatus?: ExecutionStatus
  nextRunAt?: string
  runCount: number
  failureCount: number
  createdAt: string
  updatedAt: string
}

export interface DeliveryConfig {
  email?: {
    recipients: string[]
    cc?: string[]
    subjectTemplate?: string
    bodyTemplate?: string
  }
  slack?: {
    channel: string
    webhookUrl?: string
  }
  teams?: {
    webhookUrl: string
  }
  whatsapp?: {
    phoneNumber: string
    templateName?: string
    useTextMessage?: boolean
  }
  webhook?: {
    webhookUrl: string
    headers?: Record<string, string>
    authHeader?: string
  }
  s3?: {
    bucket: string
    prefix?: string
    region?: string
  }
}

export interface ReportExecution {
  id: string
  tenantId: number
  templateId?: string
  scheduleId?: string
  executionType: 'scheduled' | 'manual' | 'api'
  status: ExecutionStatus
  startedAt: string
  completedAt?: string
  durationSeconds?: number
  reportType: ReportType
  format: ReportFormat
  dateRangeStart: string
  dateRangeEnd: string
  configUsed?: Record<string, any>
  filePath?: string
  fileSizeBytes?: number
  fileUrl?: string
  fileUrlExpiresAt?: string
  rowCount?: number
  metricsSummary?: Record<string, any>
  errorMessage?: string
  errorDetails?: Record<string, any>
  triggeredByUserId?: number
}

export interface ReportDelivery {
  id: string
  executionId: string
  channel: DeliveryChannel
  recipient: string
  status: DeliveryStatus
  sentAt?: string
  deliveredAt?: string
  errorMessage?: string
  retryCount: number
}

export interface DeliveryChannelConfig {
  id: string
  tenantId: number
  channel: DeliveryChannel
  name: string
  isActive: boolean
  isVerified: boolean
  config: Record<string, any>
  createdAt: string
  updatedAt: string
}

export interface CreateTemplateRequest {
  name: string
  description?: string
  reportType: ReportType
  config?: ReportConfig
  defaultFormat?: ReportFormat
  availableFormats?: ReportFormat[]
  templateHtml?: string
  chartConfig?: Record<string, any>
}

export interface CreateScheduleRequest {
  templateId: string
  name: string
  description?: string
  frequency: ScheduleFrequency
  timezone?: string
  dayOfWeek?: number
  dayOfMonth?: number
  hour?: number
  minute?: number
  cronExpression?: string
  formatOverride?: ReportFormat
  configOverride?: Record<string, any>
  dateRangeType?: string
  deliveryChannels: DeliveryChannel[]
  deliveryConfig: DeliveryConfig
}

export interface GenerateReportRequest {
  templateId: string
  startDate: string
  endDate: string
  format?: ReportFormat
  configOverride?: Record<string, any>
  deliverTo?: DeliveryChannel[]
  deliveryConfig?: DeliveryConfig
}

export interface ReportTypeInfo {
  type: ReportType
  name: string
  description: string
  defaultMetrics?: string[]
  availableDimensions?: string[]
  availableModels?: string[]
  sections?: string[]
}

// =============================================================================
// API Functions
// =============================================================================

export const reportingApi = {
  // Templates
  getTemplates: async (params?: { reportType?: ReportType; isActive?: boolean }) => {
    const response = await apiClient.get<ApiResponse<ReportTemplate[]>>('/reporting/templates', { params })
    return response.data.data
  },

  getTemplate: async (templateId: string) => {
    const response = await apiClient.get<ApiResponse<ReportTemplate>>(`/reporting/templates/${templateId}`)
    return response.data.data
  },

  createTemplate: async (data: CreateTemplateRequest) => {
    const response = await apiClient.post<ApiResponse<ReportTemplate>>('/reporting/templates', data)
    return response.data.data
  },

  updateTemplate: async (templateId: string, data: Partial<CreateTemplateRequest>) => {
    const response = await apiClient.patch<ApiResponse<ReportTemplate>>(`/reporting/templates/${templateId}`, data)
    return response.data.data
  },

  deleteTemplate: async (templateId: string) => {
    const response = await apiClient.delete<ApiResponse<void>>(`/reporting/templates/${templateId}`)
    return response.data
  },

  // Schedules
  getSchedules: async (params?: { isActive?: boolean; templateId?: string }) => {
    const response = await apiClient.get<ApiResponse<ScheduledReport[]>>('/reporting/schedules', { params })
    return response.data.data
  },

  getSchedule: async (scheduleId: string) => {
    const response = await apiClient.get<ApiResponse<ScheduledReport>>(`/reporting/schedules/${scheduleId}`)
    return response.data.data
  },

  createSchedule: async (data: CreateScheduleRequest) => {
    const response = await apiClient.post<ApiResponse<ScheduledReport>>('/reporting/schedules', data)
    return response.data.data
  },

  updateSchedule: async (scheduleId: string, data: Partial<CreateScheduleRequest>) => {
    const response = await apiClient.patch<ApiResponse<ScheduledReport>>(`/reporting/schedules/${scheduleId}`, data)
    return response.data.data
  },

  deleteSchedule: async (scheduleId: string) => {
    const response = await apiClient.delete<ApiResponse<void>>(`/reporting/schedules/${scheduleId}`)
    return response.data
  },

  pauseSchedule: async (scheduleId: string) => {
    const response = await apiClient.post<ApiResponse<ScheduledReport>>(`/reporting/schedules/${scheduleId}/pause`)
    return response.data.data
  },

  resumeSchedule: async (scheduleId: string) => {
    const response = await apiClient.post<ApiResponse<ScheduledReport>>(`/reporting/schedules/${scheduleId}/resume`)
    return response.data.data
  },

  runScheduleNow: async (scheduleId: string) => {
    const response = await apiClient.post<ApiResponse<ReportExecution>>(`/reporting/schedules/${scheduleId}/run-now`)
    return response.data.data
  },

  getScheduleHistory: async (scheduleId: string, limit?: number) => {
    const response = await apiClient.get<ApiResponse<ReportExecution[]>>(`/reporting/schedules/${scheduleId}/history`, {
      params: { limit },
    })
    return response.data.data
  },

  // Report Generation
  generateReport: async (data: GenerateReportRequest) => {
    const response = await apiClient.post<ApiResponse<ReportExecution>>('/reporting/generate', data)
    return response.data.data
  },

  // Executions
  getExecutions: async (params?: {
    status?: ExecutionStatus
    reportType?: ReportType
    startDate?: string
    endDate?: string
    limit?: number
    offset?: number
  }) => {
    const response = await apiClient.get<ApiResponse<ReportExecution[]>>('/reporting/executions', { params })
    return response.data.data
  },

  getExecution: async (executionId: string) => {
    const response = await apiClient.get<ApiResponse<ReportExecution>>(`/reporting/executions/${executionId}`)
    return response.data.data
  },

  downloadReport: async (executionId: string) => {
    const response = await apiClient.get<ApiResponse<{
      downloadUrl: string
      expiresAt?: string
      format: ReportFormat
      fileSizeBytes?: number
    }>>(`/reporting/executions/${executionId}/download`)
    return response.data.data
  },

  // Deliveries
  getDeliveryStatus: async (executionId: string) => {
    const response = await apiClient.get<ApiResponse<ReportDelivery[]>>(`/reporting/executions/${executionId}/deliveries`)
    return response.data.data
  },

  retryDelivery: async (deliveryId: string) => {
    const response = await apiClient.post<ApiResponse<{ success: boolean; error?: string }>>(`/reporting/deliveries/${deliveryId}/retry`)
    return response.data.data
  },

  // Delivery Channel Configs
  getDeliveryChannelConfigs: async () => {
    const response = await apiClient.get<ApiResponse<DeliveryChannelConfig[]>>('/reporting/delivery-channels')
    return response.data.data
  },

  createDeliveryChannelConfig: async (data: { channel: DeliveryChannel; name: string; config: Record<string, any> }) => {
    const response = await apiClient.post<ApiResponse<DeliveryChannelConfig>>('/reporting/delivery-channels', data)
    return response.data.data
  },

  updateDeliveryChannelConfig: async (configId: string, data: Partial<DeliveryChannelConfig>) => {
    const response = await apiClient.patch<ApiResponse<DeliveryChannelConfig>>(`/reporting/delivery-channels/${configId}`, data)
    return response.data.data
  },

  deleteDeliveryChannelConfig: async (configId: string) => {
    const response = await apiClient.delete<ApiResponse<void>>(`/reporting/delivery-channels/${configId}`)
    return response.data
  },

  verifyDeliveryChannel: async (configId: string) => {
    const response = await apiClient.post<ApiResponse<{ verified: boolean; error?: string }>>(`/reporting/delivery-channels/${configId}/verify`)
    return response.data.data
  },

  // Report Types Info
  getReportTypes: async () => {
    const response = await apiClient.get<ApiResponse<{
      reportTypes: ReportTypeInfo[]
      formats: ReportFormat[]
      frequencies: ScheduleFrequency[]
      deliveryChannels: DeliveryChannel[]
      dateRangeTypes: string[]
    }>>('/reporting/report-types')
    return response.data.data
  },
}

// =============================================================================
// React Query Hooks
// =============================================================================

// Templates
export function useReportTemplates(params?: { reportType?: ReportType; isActive?: boolean }) {
  return useQuery({
    queryKey: ['reporting', 'templates', params],
    queryFn: () => reportingApi.getTemplates(params),
  })
}

export function useReportTemplate(templateId: string) {
  return useQuery({
    queryKey: ['reporting', 'templates', templateId],
    queryFn: () => reportingApi.getTemplate(templateId),
    enabled: !!templateId,
  })
}

export function useCreateReportTemplate() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.createTemplate,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'templates'] })
    },
  })
}

export function useUpdateReportTemplate() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ templateId, data }: { templateId: string; data: Partial<CreateTemplateRequest> }) =>
      reportingApi.updateTemplate(templateId, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'templates'] })
    },
  })
}

export function useDeleteReportTemplate() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.deleteTemplate,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'templates'] })
    },
  })
}

// Schedules
export function useReportSchedules(params?: { isActive?: boolean; templateId?: string }) {
  return useQuery({
    queryKey: ['reporting', 'schedules', params],
    queryFn: () => reportingApi.getSchedules(params),
  })
}

export function useReportSchedule(scheduleId: string) {
  return useQuery({
    queryKey: ['reporting', 'schedules', scheduleId],
    queryFn: () => reportingApi.getSchedule(scheduleId),
    enabled: !!scheduleId,
  })
}

export function useCreateReportSchedule() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.createSchedule,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'schedules'] })
    },
  })
}

export function useUpdateReportSchedule() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ scheduleId, data }: { scheduleId: string; data: Partial<CreateScheduleRequest> }) =>
      reportingApi.updateSchedule(scheduleId, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'schedules'] })
    },
  })
}

export function useDeleteReportSchedule() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.deleteSchedule,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'schedules'] })
    },
  })
}

export function usePauseReportSchedule() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.pauseSchedule,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'schedules'] })
    },
  })
}

export function useResumeReportSchedule() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.resumeSchedule,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'schedules'] })
    },
  })
}

export function useRunScheduleNow() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.runScheduleNow,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'executions'] })
      queryClient.invalidateQueries({ queryKey: ['reporting', 'schedules'] })
    },
  })
}

export function useScheduleHistory(scheduleId: string, limit?: number) {
  return useQuery({
    queryKey: ['reporting', 'schedules', scheduleId, 'history', limit],
    queryFn: () => reportingApi.getScheduleHistory(scheduleId, limit),
    enabled: !!scheduleId,
  })
}

// Report Generation
export function useGenerateReport() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.generateReport,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'executions'] })
    },
  })
}

// Executions
export function useReportExecutions(params?: {
  status?: ExecutionStatus
  reportType?: ReportType
  startDate?: string
  endDate?: string
  limit?: number
  offset?: number
}) {
  return useQuery({
    queryKey: ['reporting', 'executions', params],
    queryFn: () => reportingApi.getExecutions(params),
  })
}

export function useReportExecution(executionId: string) {
  return useQuery({
    queryKey: ['reporting', 'executions', executionId],
    queryFn: () => reportingApi.getExecution(executionId),
    enabled: !!executionId,
  })
}

export function useDownloadReport(executionId: string) {
  return useQuery({
    queryKey: ['reporting', 'executions', executionId, 'download'],
    queryFn: () => reportingApi.downloadReport(executionId),
    enabled: !!executionId,
  })
}

// Deliveries
export function useDeliveryStatus(executionId: string) {
  return useQuery({
    queryKey: ['reporting', 'executions', executionId, 'deliveries'],
    queryFn: () => reportingApi.getDeliveryStatus(executionId),
    enabled: !!executionId,
  })
}

export function useRetryDelivery() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.retryDelivery,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'executions'] })
    },
  })
}

// Delivery Channel Configs
export function useDeliveryChannelConfigs() {
  return useQuery({
    queryKey: ['reporting', 'delivery-channels'],
    queryFn: reportingApi.getDeliveryChannelConfigs,
  })
}

export function useCreateDeliveryChannelConfig() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.createDeliveryChannelConfig,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'delivery-channels'] })
    },
  })
}

export function useUpdateDeliveryChannelConfig() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ configId, data }: { configId: string; data: Partial<DeliveryChannelConfig> }) =>
      reportingApi.updateDeliveryChannelConfig(configId, data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'delivery-channels'] })
    },
  })
}

export function useDeleteDeliveryChannelConfig() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.deleteDeliveryChannelConfig,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'delivery-channels'] })
    },
  })
}

export function useVerifyDeliveryChannel() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: reportingApi.verifyDeliveryChannel,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['reporting', 'delivery-channels'] })
    },
  })
}

// Report Types Info
export function useReportTypes() {
  return useQuery({
    queryKey: ['reporting', 'report-types'],
    queryFn: reportingApi.getReportTypes,
    staleTime: Infinity, // This data rarely changes
  })
}
